lib/Validation.groovy

```diff
@@ -0,0 +1,132 @@
+    */
+     private static boolean validateParamPair(given_param, json_param, log){
+        def param_type = json_param.value['type']
+        def valid_param = false
+        def required = json_param.value['required']
+        def param_enum = json_param.value['enum']
+        // Get the expected class
+        def schema_param_class = determineDefaultClass(json_param)
+
+        
+        // Check only if required or parameter is given
+        if (required || given_param){
+            def given_param_class = given_param.getClass()
+
+                switch(param_type) {
+                    case 'string':
+                        // memory
+                        if (schema_param_class == nextflow.util.MemoryUnit || given_param_class == nextflow.util.MemoryUnit){
+                            valid_param = given_param.toString() ==~ /\d+\.?\s*[KMGT]?B?/
+                        }
+                         // duration
+                        else if (schema_param_class == nextflow.util.Duration || given_param_class == nextflow.util.Duration){
+                            valid_param = given_param.toString() ==~ /\d+\.?\s*[mhd]/
+                        }
+                         // hashmap
+                        else if (schema_param_class == LinkedHashMap || given_param_class == LinkedHashMap){
+                            valid_param = true // TODO change this later too
+                        }
+                         // enum
+                        else if (param_enum){
+                            valid_param = param_enum.contains(given_param)
+                        }
+                        // normal string
+                        else {
+                            valid_param = given_param_class == String
+                        }  
+                        break
+                    case 'boolean':
+                        if (given_param_class == Boolean){
+                            valid_param = true
+                        }
+                        else if (given_param){
+                            valid_param = true
+                        }
+                        break
+                    case 'integer':
+                        valid_param = given_param_class == Integer
+                        break
+                    case 'number':
+                        valid_param = given_param_class == BigDecimal
+                        break
+                }
+
+            if (!valid_param){
+                log.error "ERROR: Parameter ${json_param.key} is wrong type! Expected ${schema_param_class}, found ${param_type}, ${given_param}"
+                if (param_enum){
+                    log.error "Must be one of: ${param_enum}"
+                }
+            }
+
+        }
+        return valid_param
+     }
+
+    private static Class determineDefaultClass(json_param){
+        def default_value = json_param.value['default']
+        def schema_param_class = default_value.getClass()
```

Comment:
Any reason not to use the `type` that is set in the schema? eg. https://github.com/nf-core/cageseq/blob/0e58db8cce0555684fb71ccaa8ced7c7a8fa4fdc/nextflow_schema.json#L38

Reply:
not anymore ^^